src/librustc_typeck/check/coercion.rs

```diff
@@ -555,9 +555,19 @@ impl<'f, 'gcx, 'tcx> Coerce<'f, 'gcx, 'tcx> {
                 ty::Predicate::Trait(ref tr) if traits.contains(&tr.def_id()) => {
                     if unsize_did == tr.def_id() {
                         let sty = &tr.skip_binder().input_types().nth(1).unwrap().sty;
-                        if let ty::Tuple(..) = sty {
-                            debug!("coerce_unsized: found unsized tuple coercion");
-                            has_unsized_tuple_coercion = true;
+                        match sty {
+                            ty::Tuple(..) => {
+                                debug!("coerce_unsized: found unsized tuple coercion");
+                                has_unsized_tuple_coercion = true;
+                            }
+                            ty::Dynamic(data, _) => {
+                                if self.tcx.features().object_safe_for_dispatch {
+                                    self.check_coercion_object_safe(&mut selcx,
```

Comment:
Ok so deep diving into my memory, I think what I was getting at was trying to turn change the error message: 

```rust
 LL |     Box::new(S) as Box<dyn Trait>;
    |     ^^^^^^^^^^^ the trait `Trait` cannot be made into an object
    |
    = note: the trait cannot require that `Self : Sized`
    = note: required because of the requirements on the impl of `std::ops::CoerceUnsized<std::boxed::Box<dyn Trait>>` for `std::boxed::Box<S>`
```
into the following:

```rust
 LL |     Box::new(S) as Box<dyn Trait>;
    |                    ^^^^^^^^^^^^^^ the trait `Trait` cannot be made into an object
    |
    = note: the trait cannot require that `Self : Sized`
    = note: required by cast to type 'std::boxed::Box<dyn Trait>'
```

I though that trying to coerce a struct into a non-object-safe trait will be quite a common occurance, and the first error message will serve as a strong deterrent to sanity for new adopters. I think the new error message is better for two reasons:

1. Obviously, there is no mention of CoerceUnsized.
2. The error span marks the 'trying to construct a non-object safe trait object' part of the code.

I think that question 1 could probably be resolved by reworking my diagnostics commit and getting the information from trait selection instead, but I'm not sure if the desired span survives is available at that point. Will have to do some more digging for that.